File: C/PathPlanner.c
/**
 * pseudo code
 *
 * 1. Reduce input waypoints to only essential (direction changing) node points
 * 2. Calculate how many total datapoints we need to satisfy the controller for "playback"
 * 3. Simultaneously inject and smooth the path until we end up with a smooth path with required number
 *    of datapoints, and which follows the waypoint path.
 * numWayPoints - number of waypoints
 * dimWayPoint - dimensions of the waypoints
 * maxTtotalTimeime - the max time allowed completing the path
 * timeStep - the frequency of the motor controller
 */

#include <stdlib.h>
#include <stdio.h>


/**
 * This method calculates the optimal parameters for determining what amount of nodes to inject into the path
 * to meet the time restraint. This approach uses an iterative process to inject and smooth, yielding more desirable
 * results for the final smooth path.
 * Big O: Constant Time
 */
int* injectionCounter2Steps(float numNodeOnlyPoints, float maxTimeToComplete, float timeStep)
{
    int first = 0;
    int second = 0;
    int third = 0;

    float oldPointsTotal = 0;

    int* ret = malloc(3*sizeof(int));

    float totalPoints = maxTimeToComplete / timeStep;

    if (totalPoints < 100)
    {
        float pointsFirst;
        float pointsTotal;


        for (int i = 4; i <= 6; i++) {
            for (int j = 1; j <= 8; j++) {
                pointsFirst = i * (numNodeOnlyPoints - 1) + numNodeOnlyPoints;
                pointsTotal = (j * (pointsFirst - 1) + pointsFirst);

                if (pointsTotal <= totalPoints && pointsTotal > oldPointsTotal)
                {
                    first = i;
                    second = j;
                    oldPointsTotal = pointsTotal;
                }
            }
        }
    } else {
        float pointsFirst;
        float pointsSecond;
        float pointsTotal;

        for (int i = 1; i <= 5; i++) {
            for (int j = 1; j <= 8; j++) {
                for (int k = 1; k < 8; k++) {
                    pointsFirst = i * (numNodeOnlyPoints - 1) + numNodeOnlyPoints;
                    pointsSecond = (j * (pointsFirst - 1) + pointsFirst);
                    pointsTotal = (k * (pointsSecond - 1) + pointsSecond);

                    if (pointsTotal <= totalPoints) {
                        first = i;
                        second = j;
                        third = k;
                    }
                }
            }
        }
    }

    ret[0] = first;
    ret[1] = second;
    ret[2] = third;

    return ret;
}

/**
 * Method upsamples the Path by linear injection. The result providing more waypoints along the path.
 * BigO: Order N * injection#
 */
float** inject(int numWayPoints, int dimWayPoint, float** orig, int numToInject) {
    //create extended n Dimensional array to hold additional points
    int newNumWayPoints = numWayPoints + ((numToInject) * (numWayPoints- 1));
    float** morePoints = malloc(newNumWayPoints * sizeof(float*));
    for (int i = 0; i < newNumWayPoints; i++) {
      morePoints[i] = malloc(dimWayPoint * sizeof(float));
    }

    int index = 0;

    //loop through original array
    for (int i = 0; i < numWayPoints - 1; i++) {
        //copy first

        for(int j = 0; j < dimWayPoint; j++) {
            morePoints[index][j] = orig[i][j];
        }

        index++;

        for (int j = 1; j < numToInject + 1; j++) {
            for (int k = 0; k < dimWayPoint; k++) {
                //calculate intermediate x points between j and j+1 original points
                morePoints[index][k] = j * ((orig[i + 1][k] - orig[i][k]) / (numToInject + 1)) + orig[i][k];
            }
            index++;
        }
    }

    //copy last
    for(int j = 0; j < dimWayPoint; j++) {
        morePoints[index][j] = orig[numWayPoints-1][j];
    }

    return morePoints;
}

/**
 * Optimization algorithm, which optimizes the data points in path to create a smooth trajectory.
 * This optimization uses gradient descent. While unlikely, it is possible for this algorithm to never
 * converge. If this happens, try increasing the tolerance level.
 * BigO: N^x, where X is the number of of times the while loop iterates before tolerance is met.
 */
float** smoother(int numWayPoints, int dimWayPoint, float** path, float weight_data, float weight_smooth, float tolerance) {
    //copy array
    float**  newPath = malloc(numWayPoints * sizeof(float*));
    for (int i = 0; i < numWayPoints; i++) {
        newPath[i] = malloc(dimWayPoint * sizeof(float));
    }

    for(int i = 0; i < numWayPoints; i++) {
        for(int j = 0; j < dimWayPoint; j++) {
            newPath[i][j] = path[i][j];
        }
    }

    double change = tolerance;
    while (change >= tolerance) {
        change = 0.0;
        for (int i = 1; i < numWayPoints - 1; i++) {
            for (int j = 0; j < dimWayPoint; j++) {
                double aux = newPath[i][j];
                newPath[i][j] += weight_data * (path[i][j] - newPath[i][j]) + weight_smooth * (newPath[i - 1][j] + newPath[i + 1][j] - (2.0 * newPath[i][j]));
                change += abs(aux - newPath[i][j]);
            }
        }
    }

    return newPath;
}


float** smoothPath(int numWayPoints, int dimWayPoint, float** wayPoints, float totalTime, float timeStep)
{
    //the path we are going to return
    float** path;

    //important values used for smoothing
    float pathAlpha = 0.7;
    float pathBeta = 0.3;
    float pathTolerance = 0.0000001;

    int numPoints = numWayPoints;
    int* injections = injectionCounter2Steps(numWayPoints, totalTime, timeStep);
    //iteratively inject and smooth the path
    for (int i = 0; i < 3; i++) {
        if (i == 0) {
            path = inject(numPoints, dimWayPoint, wayPoints, injections[0]);

            //update the number of points in our path before proceeding
            numPoints = numPoints + ((injections[i]) * (numPoints- 1));

            path = smoother(numPoints, dimWayPoint, path, pathAlpha, pathBeta, pathTolerance);
        } else {
            path = inject(numPoints, dimWayPoint, path, injections[i]);

            //update the number of points in our path before proceeding
            numPoints = numPoints + ((injections[i]) * (numPoints- 1));

            path = smoother(numPoints, dimWayPoint, path, pathAlpha, pathBeta, pathTolerance);
        }
    }
    //shift all the points down 1 and store the length in [0][0], including
    //the added point.
    float** finalPath = malloc((numPoints + 1) * sizeof(float*));
    for(int i = 0; i < numPoints + 1; i++) {
      finalPath[i] = malloc(dimWayPoint * sizeof(float));
    }

    finalPath[0][0] = numPoints + 1;

    for(int i = 0; i < numPoints; i++) {
        for(int j = 0; j < dimWayPoint; j++) {
            finalPath[i+1][j] = path[i][j];
        }
    }

    //clean up memory
    for (int i = 0; i < numPoints; i++) {
      free(path[i]);
    }
    free(path);

    return finalPath;
}

void test()
{
    int num_points = 5;
    int dim_points = 2;
    float** points = malloc(num_points * sizeof(float*));
    for (int i = 0; i < num_points; i++) {
      points[i] = malloc(dim_points * sizeof(float));
    }

    points[0][0] = 1;
    points[0][1] = 2;
    points[1][0] = 2;
    points[1][1] = 7;
    points[2][0] = 4;
    points[2][1] = 7;
    points[3][0] = 6;
    points[3][1] = 9;
    points[4][0] = 10;
    points[4][1] = 11;

    float totalTime = 15;
    float timeStep = .1;

    float** sPath;
    sPath = smoothPath(num_points, dim_points, points, totalTime, timeStep);

    for(int i = 1; i < sPath[0][0]; i++) {
        for(int j = 0; j < dim_points; j++) {
            printf("%f ", sPath[i][j]);
        }
        printf("\n");
    }
    //clean up memory
    for (int i = 0; i < num_points; i++) {
      free(points[i]);
    }
    free(points);

    for (int i = 0; i < sPath[9][0]; i++) {
      free(sPath[i]);
    }
    free(sPath);
}

int main()
{
    test();
    return 0;
}

File: Python/PathPlanner.py
from copy import deepcopy


#
# This method calculates the optimal parameters for determining what amount of nodes to inject into the path
# to meet the time restraint. This approach uses an iterative process to inject and smooth, yielding more desirable
# results for the final smooth path.
# Big O: Constant Time
def injection_counter_to_steps(path_length, max_time_to_complete_, time_step):
    first = 0
    second = 0
    third = 0

    old_points_total = 0

    ret = []

    total_points = max_time_to_complete_ / time_step

    if total_points < 100:
        points_first = None
        points_total = None
        for i in range(4, 6):
            for j in range(1, 8):
                points_first = i * (path_length - 1) + path_length
                points_total = (j * (points_first - 1) + points_first)

                if old_points_total < points_total <= total_points:
                    first = i
                    second = j
                    old_points_total = points_total
    else:
        points_first = None
        points_second = None
        points_total = None
        for i in range(1, 6):
            for j in range(1, 7):
                for k in range(1, 8):
                    points_first = i * (path_length - 1) + path_length;
                    points_second = (j * (points_first - 1) + points_first);
                    points_total = (k * (points_second - 1) + points_second);

                    if points_total <= total_points:
                        first = i
                        second = j
                        third = k

    return [first, second, third]


# Method up-samples the Path by linear injection. The result providing more waypoints along the path.
# BigO: Order N * injection
def inject(orig, num_to_inject):
    more_points = []
    length = len(orig) + (num_to_inject * (len(orig) - 1))
    # loop through original array
    for i in range(0, length):
        tmp = []
        for dim in orig[0]:
            tmp.append(0)
        more_points.append(tmp)

    index = 0

    for i in range(0, len(orig) - 1):
        for j in range(0, len(orig[i])):
            more_points[index][j] = orig[i][j]

        index += 1

        for j in range(1, num_to_inject + 1):
            for k in range(0, len(orig[0])):
                # calculate intermediate x points between j and j+1 original points
                more_points[index][k] = j * ((orig[i + 1][k] - orig[i][k]) / (num_to_inject + 1)) + orig[i][k]
            index += 1

    # copy last
    for j in range(0, len(orig[-1])):
        more_points[-1][j] = orig[-1][j]

    return more_points


def smoother(orig, weight_data, weight_smooth, tolerance):
        # copy path
        new_path = deepcopy(orig)

        change = tolerance
        while change >= tolerance:
            change = 0.0
            for i in range(0, len(new_path) - 1):
                for j in range(0, len(orig[i])):
                    aux = new_path[i][j]
                    new_path[i][j] += weight_data * (orig[i][j] - new_path[i][j]) + weight_smooth * (new_path[i - 1][j] + new_path[i + 1][j] - (2.0 * new_path[i][j]))
                    change += abs(aux-new_path[i][j])
        return new_path


def smooth_path(orig, total_time, time_step):
    # smoothing parameters
    path_alpha = 0.7
    path_beta = 0.3
    path_tolerance = 0.0000001

    # smooth path that will be returned
    path = []
    injections = injection_counter_to_steps(len(orig), total_time, time_step)
    for i in injections:
        if injections.index(i) == 0:
            path = inject(orig, i)
            path = smoother(path, path_alpha, path_beta, path_tolerance)
        else:
            path = inject(path, i)
            path = smoother(path, path_alpha, path_beta, path_tolerance)

    return path


def test():
    points = []

    point1 = [1, 2]
    point2 = [2, 7]
    point3 = [4, 7]
    point4 = [6, 9]
    point5 = [10, 11]

    points.append(point1)
    points.append(point2)
    points.append(point3)
    points.append(point4)
    points.append(point5)

    total_time = 15
    time_step = .1
    path = smooth_path(points, total_time, time_step)

    for point in path:
        print(point)


if __name__ == "__main__":
    test()
